File: simple-stack-example-mvvm-fragments/src/main/java/com/zhuinden/simplestackexamplemvvm/core/database/QueryBuilder.java
package com.zhuinden.simplestackexamplemvvm.core.database;

import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

public class QueryBuilder {
    private QueryBuilder() {
    }

    public enum Sort {
        NONE(""),
        ASC("ASC"),
        DESC("DESC");

        private final String value;

        private Sort(String value) {
            this.value = value;
        }
    }

    private boolean distinct = false;

    private String tableName;

    private String[] columns;

    private String whereCondition;

    private String[] whereArgs;

    private String groupBy;

    private String having;

    private String orderBy;

    private Sort sortOrder;

    private String limit = null;

    public QueryBuilder distinct(boolean distinct) {
        this.distinct = true;
        return this;
    }

    public QueryBuilder tableName(String tableName) {
        this.tableName = tableName;
        return this;
    }

    public QueryBuilder columns(String[] columns) {
        this.columns = columns;
        return this;
    }

    public QueryBuilder where(String condition, Object... args) {
        this.whereCondition = condition;
        int length = args == null ? 0 : args.length;
        String[] selectionArgs = new String[length];
        for(int i = 0; i < length; i++) {
            Object arg = args[i];
            if(arg instanceof String) {
                selectionArgs[i] = (String) arg;
            } else {
                selectionArgs[i] = arg.toString();
            }
        }
        this.whereArgs = selectionArgs;
        return this;
    }

    public QueryBuilder groupBy(String groupBy) {
        this.groupBy = groupBy;
        return this;
    }

    public QueryBuilder having(String having) {
        this.having = having;
        return this;
    }

    public QueryBuilder orderBy(String orderBy, Sort sortOrder) {
        this.orderBy = orderBy;
        this.sortOrder = sortOrder;
        return this;
    }

    public QueryBuilder limit(Integer limit) {
        if(limit == null) {
            this.limit = null;
        } else {
            this.limit = String.valueOf(limit);
        }
        return this;
    }

    public static QueryBuilder of(DatabaseManager.Table table) {
        QueryBuilder queryBuilder = new QueryBuilder();
        queryBuilder.tableName = table.getTableName();
        queryBuilder.columns = table.getAllQueryFields();
        return queryBuilder;
    }

    public DatabaseManager.QueryDefinition buildDefinition() {
        return (database, table) -> executeQuery(database);
    }

    public Cursor executeQuery(SQLiteDatabase database) {
        return database.query(
                distinct,
                tableName,
                columns,
                whereCondition,
                whereArgs,
                groupBy,
                having,
                orderBy == null ? null : orderBy + " " + sortOrder.value,
                limit
        );
    }
}

File: samples/advanced-samples/mvvm-sample/src/main/java/com/zhuinden/simplestackexamplemvvm/core/database/QueryBuilder.kt
package com.zhuinden.simplestackexamplemvvm.core.database


import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import com.zhuinden.simplestackexamplemvvm.core.database.DatabaseManager.QueryDefinition

/**
 * Created by Zhuinden on 2017.07.26..
 */
class QueryBuilder private constructor() {
    enum class Sort(val value: String) {
        NONE(""), ASC("ASC"), DESC("DESC");
    }

    private var distinct = false
    private var tableName: String? = null
    private var columns: Array<String>? = null
    private var whereCondition: String? = null
    private var whereArgs: Array<String>? = null
    private var groupBy: String? = null
    private var having: String? = null
    private var orderBy: String? = null
    private var sortOrder: Sort? = null
    private var limit: String? = null

    fun distinct(distinct: Boolean): QueryBuilder {
        this.distinct = distinct
        return this
    }

    fun tableName(tableName: String): QueryBuilder {
        this.tableName = tableName
        return this
    }

    fun columns(columns: Array<String>): QueryBuilder {
        this.columns = columns
        return this
    }

    fun where(condition: String, vararg args: Any): QueryBuilder {
        whereCondition = condition
        val length = args.size ?: 0
        val selectionArgs = arrayOfNulls<String>(length)
        for (i in 0 until length) {
            val arg = args[i]
            if (arg is String) {
                selectionArgs[i] = arg
            } else {
                selectionArgs[i] = arg.toString()
            }
        }
        whereArgs = selectionArgs.mapNotNull { it }.toTypedArray()
        return this
    }

    fun groupBy(groupBy: String?): QueryBuilder {
        this.groupBy = groupBy
        return this
    }

    fun having(having: String?): QueryBuilder {
        this.having = having
        return this
    }

    fun orderBy(orderBy: String?, sortOrder: Sort?): QueryBuilder {
        this.orderBy = orderBy
        this.sortOrder = sortOrder
        return this
    }

    fun limit(limit: Int?): QueryBuilder {
        if (limit == null) {
            this.limit = null
        } else {
            this.limit = limit.toString()
        }
        return this
    }

    fun buildDefinition(): QueryDefinition {
        return QueryDefinition { database: SQLiteDatabase, table: DatabaseManager.Table? -> executeQuery(database) }
    }

    fun executeQuery(database: SQLiteDatabase): Cursor {
        return database.query(
            distinct,
            tableName,
            columns,
            whereCondition,
            whereArgs,
            groupBy,
            having,
            if (orderBy == null) null else orderBy + " " + sortOrder!!.value,
            limit
        )
    }

    companion object {
        fun of(table: DatabaseManager.Table): QueryBuilder {
            val queryBuilder = QueryBuilder()
            queryBuilder.tableName = table.tableName
            queryBuilder.columns = table.allQueryFields
            return queryBuilder
        }
    }
}